src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -1715,35 +1715,52 @@ abstract class RefChecks extends Transform {
     }
 
     /** Check that boolean literals are passed as named args.
-     *  The rule is enforced when the type of the parameter is `Boolean`.
-     *  The rule is relaxed when the method has exactly one boolean parameter
+     *  The rule is enforced when the type of the parameter is `Boolean`,
+     *  and there is more than one parameter with an unnamed argument.
+     *  The stricter internal lint warns for any unnamed argument,
+     *  except that the rule is relaxed when the method has exactly one boolean parameter
      *  and it is the first parameter, such as `assert(false, msg)`.
      */
     private def checkNamedBooleanArgs(fn: Tree, args: List[Tree]): Unit = {
       val sym = fn.symbol
       def applyDepth: Int = {
         def loop(t: Tree, d: Int): Int =
           t match {
-            case Apply(f, _) => loop(f, d+1)
+            case Apply(t, _) => loop(t, d+1)
             case _ => d
           }
         loop(fn, 0)
       }
-      def isAssertParadigm(params: List[Symbol]): Boolean = !sym.isConstructor && !sym.isCaseApplyOrUnapply && {
-        params match {
-          case h :: t => h.tpe == BooleanTpe && !t.exists(_.tpe == BooleanTpe)
-          case _ => false
-        }
-      }
-      if (settings.lintNamedBooleans && !sym.isJavaDefined && !args.isEmpty) {
+      if (settings.warnNamedLiteral.value && !sym.isJavaDefined && !args.isEmpty) {
+        def isUnnamedArg(t: Tree) = t.hasAttachment[UnnamedArg.type]
         val params = sym.paramLists(applyDepth)
-        if (!isAssertParadigm(params))
-          foreach2(args, params)((arg, param) => arg match {
-            case Literal(Constant(_: Boolean))
-            if arg.hasAttachment[UnnamedArg.type] && param.tpe.typeSymbol == BooleanClass && !param.deprecatedParamName.contains(nme.NO_NAME) =>
-              runReporting.warning(arg.pos, s"Boolean literals should be passed using named argument syntax for parameter ${param.name}.", WarningCategory.LintNamedBooleans, sym)
-            case _ =>
-          })
+        val numBools = params.count(_.tpe == BooleanTpe)
+        def onlyLeadingBool = numBools == 1 && params.head.tpe == BooleanTpe
+        val checkable = if (settings.Ylint.value) numBools > 0 && !onlyLeadingBool else numBools >= 2
+        if (checkable) {
+          val (unnamed, numSuspicious) = args.lazyZip(params).iterator
+            .foldLeft((List.empty[(Tree, Symbol)], 0)) { (acc, ap) =>
+              ap match {
+                case (arg, param) if param.tpe.typeSymbol == BooleanClass && !param.deprecatedParamName.contains(nme.NO_NAME) =>
+                  arg match {
+                    case Literal(Constant(_: Boolean)) =>
+                      if (isUnnamedArg(arg)) (ap :: acc._1, acc._2 + 1) else acc
+                    case _ =>
+                      if (arg.symbol != null && arg.symbol.isDefaultGetter) (acc._1, acc._2 + 1) else acc
```

Comment:
`isDefaultGetter` doesn't work when the named args tranformation did its thing. Maybe an unlikely corner case though.

```scala
scala> def bul = true
def bul: Boolean

scala> def f(n: Int, inv: Boolean = false, up: Boolean = false, down: Boolean = false) = 0
def f(n: Int, inv: Boolean, up: Boolean, down: Boolean): Int

scala> f(1, true, up = bul)
            ^
       warning: Boolean literals should be passed using named argument syntax for parameter inv. [quickfixable]
val res0: Int = 0

scala> f(1, true, down = bul)
val res1: Int = 0
```

```scala
      private[this] val res2: Int = {
        <artifact> val x$1: Int(1) = 1;
        <artifact> val x$2: Boolean(true) = true;
        <artifact> val x$3: Boolean = $read.this.$line3$read.$iw.bul;
        <artifact> val x$4: Boolean = $read.this.$line4$read.$iw.f$default$3;
        $read.this.$line4$read.$iw.f(1, true, x$4, x$3)
      };
```

Reply:
The unlikely corner case complains the loudest.

Reply:
thanks, this is another instance of nothing good happens after midnight. I was looking at this, but a dumb bug in my refactor used up the rest of my midnight oil. It's an edge case because it requires many boolean parameters and one boolean literal arg.

It would be nice if default arg were an attachment instead of a name test. A name can be long and linting prefers efficient up-front tests even if emitting diagnostics is expensive.